src/generator.cpp

```
@@ -1285,6 +1313,649 @@ void Generator::GeneratorImpl::processEquationAst(const GeneratorEquationAstPtr
+                }
+            }
+
+            // Multiply, Divide: add mappings, no interest in unit compatibility.
+            if (isMultiplicativeOperator(ast)) {
+                UnitsMap newMapping;
+                if (ast->mType == GeneratorEquationAst::Type::TIMES) {
+                    newMapping = addMappings(leftMap, rightMap, 1);
+                } else {
+                    newMapping = addMappings(leftMap, rightMap, -1);
+                }
+                return newMapping;
+            }
+
+            // Checks for exponential operators, multiplies unit mappings with power
+            if (isExponentOperator(ast)) {
+                double power = 0.0;
+                if (ast->mType == GeneratorEquationAst::Type::POWER) {
+                    power = getPower(ast->mRight);
+                } else {
+                    if (ast->mLeft->mType == GeneratorEquationAst::Type::DEGREE) {
+                        power = getPower(ast->mLeft);
+                    } else {
+                        power = getPower(ast->mRight);
+                    }
+                }
+
+                //double power = getPower(ast->mRight);
+                bool correctUnits = false;
+                if (power == 0.0 && ast->mRight != nullptr && ast->mType == GeneratorEquationAst::Type::POWER) {
+                    correctUnits = isDimensionless(leftMap) && isDimensionless(rightMap); // If we have a variable as our power both the power and the quantity it is being applied to must be dimensionless
+                } else if (ast->mLeft->mType == GeneratorEquationAst::Type::DEGREE) {
+                    correctUnits = isDimensionless(leftMap);
+                    leftMap = rightMap;
+                } else {
+                    correctUnits = isDimensionless(rightMap); // Otherwise we just check the power for dimensionlessness
+                }
+
+                // If we have a square root operation
+                if (power == 0.0 && ast->mRight == nullptr) {
```

I don't quite follow this bit?

---

There are instances in the AST where a root number is not specified as a child (i.e, the root node will only have a variable child and no second child giving a root number - it happens a couple of times in some of the bigger models in the unit tests for the `validator`). So, if there is no child (a power cannot be found, so 0.0 is returned), the assumption is that the intended operation is a square root operation.

---

MathML defines a default square-root: https://www.w3.org/TR/REC-MathML/chap4_4.html#sec4.4.3.11

---

I'm not sure we should be making any assumptions anywhere?  Isn't that the point of an unambiguous model description? Rather return an error ...

---

I think a few of the models used for testing (not the ones I added in) would spontaneously combust if I was to just return an error. Perhaps when the AST gets moved out/exposed then some discussion can be made as to whether default roots like ```sqrt``` can be made into their own type of internal node.

---

the definition of the root operator is very clear in this regard. It is not an assumption. A root with two arguments has the degree (number) and the thing it is applied to; a root with a single argument has degree 2. (sorry if I'm missing something, trying to be helpful without reading everything ;-)